File: python/bench.py
import argparse
import json
import multiprocessing as mp
import sys
import threading
import time


def rss_bytes():
    try:
        import psutil
        return psutil.Process().memory_info().rss
    except:
        import resource
        return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss * 1024


def splitmix64(seed):
    x = seed & ((1 << 64) - 1)
    while True:
        x = (x + 0x9E3779B97F4A7C15) & ((1 << 64) - 1)
        z = x
        z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9 & ((1 << 64) - 1)
        z = (z ^ (z >> 27)) * 0x94D049BB133111EB & ((1 << 64) - 1)
        z = z ^ (z >> 31)
        yield z


def worker_thread(shards, locks, keys, per, read_ratio, seed, tid):
    rnd = splitmix64(seed + tid)
    reads = int(read_ratio * 1000)
    for _ in range(per):
        k = int(next(rnd) % keys)
        r = int(next(rnd) % 1000)
        s = k % len(shards)
        if r < reads:
            with locks[s]:
                _ = shards[s].get(k, 0)
        else:
            with locks[s]:
                shards[s][k] = shards[s].get(k, 0) + 1


def worker_process(args):
    (keys, per, read_ratio, seed, pid, shardsN) = args
    shards = [dict() for _ in range(shardsN)]
    # prefill
    for i in range(keys):
        shards[i % shardsN][i] = 0
    rnd = splitmix64(seed + pid)
    reads = int(read_ratio * 1000)
    for _ in range(per):
        k = int(next(rnd) % keys)
        r = int(next(rnd) % 1000)
        s = k % shardsN
        if r < reads:
            _ = shards[s].get(k, 0)
        else:
            shards[s][k] = shards[s].get(k, 0) + 1
    return 0


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--threads", type=int, default=8)
    ap.add_argument("--iterations", type=int, default=2_000_000)
    ap.add_argument("--keys", type=int, default=100_000)
    ap.add_argument("--read-ratio", type=float, default=0.9)
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--model", type=str, default="threads")
    ap.add_argument("--shards", type=int, default=64)
    args = ap.parse_args()

    per = args.iterations // args.threads
    t0 = time.time()

    if args.model == "threads":
        shards = [dict() for _ in range(args.shards)]
        locks = [threading.Lock() for _ in range(args.shards)]
        for i in range(args.keys):
            shards[i % args.shards][i] = 0
        ts = []
        for i in range(args.threads):
            t = threading.Thread(target=worker_thread,
                                 args=(shards, locks, args.keys, per, args.read_ratio, args.seed, i))
            t.start();
            ts.append(t)
        for t in ts: t.join()
    elif args.model == "processes":
        with mp.Pool(processes=args.threads) as pool:
            pool.map(worker_process,
                     [(args.keys, per, args.read_ratio, args.seed, i, args.shards) for i in range(args.threads)])
    else:
        print("unknown model", file=sys.stderr);
        sys.exit(1)

    dur = int((time.time() - t0) * 1000)
    res = {
        "runtime": f"python{sys.version.split()[0]}",
        "model": args.model,
        "threads": args.threads,
        "iterations": per * args.threads,
        "keys": args.keys,
        "read_ratio": args.read_ratio,
        "seed": args.seed,
        "duration_ms": dur,
        "rss_bytes": rss_bytes()
    }
    print(json.dumps(res))


if __name__ == "__main__":
    main()

File: go/bench.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"math/rand/v2"
	"os"
	"runtime"
	"runtime/debug"
	"sync"
	"time"
)

type Result struct {
	Runtime    string  `json:"runtime"`
	Model      string  `json:"model"`
	Threads    int     `json:"threads"`
	Iterations int     `json:"iterations"`
	Keys       int     `json:"keys"`
	ReadRatio  float64 `json:"read_ratio"`
	Seed       uint64  `json:"seed"`
	DurationMS int64   `json:"duration_ms"`
	RSSBytes   uint64  `json:"rss_bytes"`
}

func rss() uint64 {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	return m.Sys // close to RSS; for external RSS, parse /proc/self/statm on Linux
}

func main() {
	threads := flag.Int("threads", 8, "")
	iters := flag.Int("iterations", 2_000_000, "")
	keys := flag.Int("keys", 100_000, "")
	readRatio := flag.Float64("read-ratio", 0.9, "")
	seed := flag.Uint64("seed", 42, "")
	model := flag.String("model", "syncmap", "syncmap|sharded")
	shardsN := flag.Int("shards", 64, "for sharded model")
	flag.Parse()

	debug.SetGCPercent(100)
	runtime.GOMAXPROCS(runtime.NumCPU())

	switch *model {
	case "syncmap":
		runSyncMap(*threads, *iters, *keys, *readRatio, *seed)
	case "sharded":
		runSharded(*threads, *iters, *keys, *readRatio, *seed, *shardsN)
	default:
		fmt.Fprintln(os.Stderr, "unknown model")
		os.Exit(1)
	}
}

func runSyncMap(threads, iterations, keys int, readRatio float64, seed uint64) {
	var m sync.Map
	// prefill
	for i := 0; i < keys; i++ {
		m.Store(i, int64(0))
	}
	runWork("go"+runtime.Version(), "threads-sync.Map", threads, iterations, keys, readRatio, seed, func(ctx context.Context, tid int, n int) {
		r := rand.New(rand.NewPCG(seed+uint64(tid), seed*1315423911+uint64(tid)))
		reads := int(readRatio * 1000.0)
		for i := 0; i < n; i++ {
			k := int(r.IntN(keys))
			if int(r.IntN(1000)) < reads {
				if v, ok := m.Load(k); ok {
					_ = v.(int64)
				}
			} else {
				for {
					v, _ := m.Load(k)
					old := int64(0)
					if v != nil {
						old = v.(int64)
					}
					if m.CompareAndSwap(k, old, old+1) {
						break
					}
					// if key absent, Store will fail CAS; ensure presence
					if v == nil {
						// initialize
						m.LoadOrStore(k, int64(0))
					}
				}
			}
		}
	}, &m)
}

type shard struct {
	mu sync.Mutex
	m  map[int]int64
}

func runSharded(threads, iterations, keys int, readRatio float64, seed uint64, shardsN int) {
	shards := make([]*shard, shardsN)
	for i := range shards {
		shards[i] = &shard{m: make(map[int]int64, keys/shardsN+1)}
	}
	// prefill
	for i := 0; i < keys; i++ {
		s := shards[i%shardsN]
		s.mu.Lock()
		s.m[i] = 0
		s.mu.Unlock()
	}
	runWork("go"+runtime.Version(), "threads-sharded", threads, iterations, keys, readRatio, seed, func(ctx context.Context, tid int, n int) {
		r := rand.New(rand.NewPCG(seed+uint64(tid), seed*1315423911+uint64(tid)))
		reads := int(readRatio * 1000.0)
		for i := 0; i < n; i++ {
			k := int(r.IntN(keys))
			s := shards[k%shardsN]
			if int(r.IntN(1000)) < reads {
				s.mu.Lock()
				_ = s.m[k]
				s.mu.Unlock()
			} else {
				s.mu.Lock()
				s.m[k]++
				s.mu.Unlock()
			}
		}
	}, shards)
}

func runWork(rt, model string, threads, iterations, keys int, readRatio float64, seed uint64, worker func(context.Context, int, int), obj any) {
	per := iterations / threads
	ctx := context.Background()
	start := time.Now()
	var wg sync.WaitGroup
	wg.Add(threads)
	for t := 0; t < threads; t++ {
		tid := t
		go func() {
			defer wg.Done()
			worker(ctx, tid, per)
		}()
	}
	wg.Wait()
	dur := time.Since(start).Milliseconds()
	_ = obj // keep from optimizing away

	// force GC to stabilize mem reading
	runtime.GC()
	time.Sleep(50 * time.Millisecond)
	res := Result{
		Runtime:    rt,
		Model:      model,
		Threads:    threads,
		Iterations: per * threads,
		Keys:       keys,
		ReadRatio:  readRatio,
		Seed:       seed,
		DurationMS: dur,
		RSSBytes:   rss(),
	}
	b, _ := json.Marshal(res)
	fmt.Println(string(b))
}
